### ISSUE TITLE
dependencies: replace requests with httpx

### ISSUE THREAD

**Issue by bastimeyer** [streamlink/streamlink#4915]
Rather than a suggestion or even proposal, I'm opening this thread merely as an early discussion for a potential future replacement of Streamlink's [`requests`](https://github.com/psf/requests) dependency in favor of [`httpx`](https://github.com/encode/httpx).

`httpx` is a new-ish replacement-library for `requests` with an (almost) idential API and it offers more features like HTTP/2 for example and it (optionally) also allows writing more modern code like asynchronous HTTP requests. There were attempts in the past to provide asynchronous APIs for `requests` via [`requests-async`](https://github.com/encode/requests-async), but that project was abandoned and its authors continued with the `httpx` project.

`requests` may be depended upon by hundred thousands or even millions of projects in the Python ecosystem, but it has become very stale in terms of upgrades and features, and some of its optional dependencies like [`PySocks`](https://github.com/Anorov/PySocks) (which [we're defining directly here for some reason](https://github.com/streamlink/streamlink/blob/5.0.1/setup.cfg#L47)) have been abandoned since years.

`httpx` uses the BSD 3-clause license and has [a completely different set of dependencies](https://github.com/encode/httpx#dependencies) compared to [`requests`](https://github.com/psf/requests/blob/v2.28.1/setup.py#L61-L66), with a much better and clear distinction between the purposes and features of its own project and its dependencies, unlike `requests` and [`urllib3`](https://github.com/urllib3/urllib3), which are sometimes difficult to understand and debug. `requests` and `urllib3` are also struggling with their future 3.0.0 and 2.0.0 releases respectively due to their vast usages and thus stability guarantees, meaning that there's a ton of outdated stuff like py2 compatibility for example still in there, among other things.

A stable 1.0.0 release of `httpx` hasn't been published yet though, so even if there were a strong consensus of everyone here and even if it were easy to replace, it wouldn't make sense just yet.

Talking about ease of replacement, since both projects share an almost identical API, replacing it as the base class of Streamlink's `HTTPSession` *should* be fairly simple. However, over the years, we've added lots of compatibility bugfixes and monkeypatches to `requests` and `urllib3` ([1](https://github.com/streamlink/streamlink/blob/5.0.1/src/streamlink/plugin/api/http_session.py#L16-L80), [2](https://github.com/streamlink/streamlink/blob/5.0.1/src/streamlink/session.py#L220-L242)), which means that this would need to be checked and evaluated first.

Then there's also the part [where both projects differ](https://www.python-httpx.org/compatibility/). Most of it should be fine though and would only require a change in the `HTTPSession` class itself, and not in any actual http session usages like stream implementations or plugins, etc. For users of Streamlink, I don't think there would be any major changes in terms of features supported by `requests` directly, like TLS/SSL or proxy stuff via env vars, etc. Streamlink also packages a fork of a `requests` `FileAdapter` for the `file://` protocol, but that should be fine for the most part, and if not, should be easy to replace or properly rewrite.

Another replacement would be `requests-mock` in Streamlink's tests which would need to be figured out as well. A substitution exists in the form of [`RESPX`](https://github.com/lundberg/respx).

Packaging-wise, the first `httpx` releases should be available since 2020 in non-monolithic package repositories, e.g.:
- https://packages.debian.org/bullseye/python3-httpx
- https://archlinux.org/packages/community/any/python-httpx/
- https://src.fedoraproject.org/rpms/python-httpx
- https://packages.gentoo.org/packages/dev-python/httpx
- https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/python-modules/httpx/default.nix
- https://build.opensuse.org/package/show/devel:languages:python/python-httpx
- https://github.com/void-linux/void-packages/tree/master/srcpkgs/python3-httpx
- https://pkgsrc.se/www/py-httpx

The ones where I couldn't find `httpx` were homebrew and solus.

The main question however is how reliable httpx will be in the future. Since it appears well maintained and looks like it has some momentum in terms of development and adoption by other projects, this doesn't look bad at all.

**Comment by gravyboat** [comment]
I'd be fine with swapping over. We can say that `requests` is long established and stable but as you noted the reliance on abandoned code is always concerning. The biggest flaw I see with `httpx` is the fact it still doesn't have a stable release after 3 years though I also don't see anything blocking us from using it now if we'd like since this isn't a commercial product and we have no responsibility to deliver any sort of stability guarantees.

**Comment by L2501** [comment]
httpx in http2 mode gets blocked by cloudflare.

**Comment by bastimeyer** [comment]
Just wanted to add to the current state of requests and urllib3:
While there doesn't seem to be any progress towards requests 3.0, urllib3 2.0 has been released as alpha in November, with a stable release planned for this year. Streamlink is limited to urllib3<1.27 via requests 2.x, but if this is going to change, then there will be some incompatibilities with current implementations. The override of urllib3's default SSL ciphers (`--http-disable-dh`) is one of them, which has been removed and providing a custom `SSLContext` is now required (which should've been done anyway instead of overriding globals of third party modules).

**Comment by bastimeyer** [comment]
> There's a 3rd option you may be interested in. [jawah/niquests](https://github.com/jawah/niquests)

Sorry, I should've responded a year ago when I had my first look at it, so let me finally comment on it now. I've been assessing the requirements and complexity of changes for switching to different HTTP libs every so often here, because it's actually not trivial.

Streamlink is at the same time also in need of a huge refactor/rewrite of its stream implementations, as well as its overall stream selection mechanism, affecting all public interfaces and huge parts of its CLI. All of these major changes are somewhat intertwined with each other, and switching the underlying HTTP libs would also affect the refactoring plans, because of how Streamlink's `HTTPSession` was implemented when it was forked. That's what makes it difficult.

While `niquests` seems really great at first glance, it's unfortunately currently impossible for Streamlink to switch to it because of how you decided to fork `requests` and `urllib3`. None of your forks are packaged by any major Linux distro (yet), and your `urllib3.future` fork also replaces the `urllib3` namespace, which makes it very difficult for packagers to package. If I would replace `requests`/`urllib3` in Streamlink with `niquests`/`urllib3.future`, I would put massive pressure on all of Streamlink's packagers (it's already a huge pain having the latest version of Streamlink packaged everywhere), and I would risk making them drop Streamlink from their repos.

Then there's also the lack of support for `trio`/`anyio` for the `AsyncSession`'s runloop, which is bad for Streamlink, as it doesn't use `asyncio`:
https://github.com/jawah/niquests/blame/f224adccd89cc248ab536fe1771c2b03180529ce/docs/user/quickstart.rst#L745

In regards to the `niquests` API, I don't see any major hurdles.

So unless the project matures and gets packaged in major Linux distros, Streamlink won't be able to switch to it, despite of all the problems it would solve and all the features it would offer.

**Comment by Ousret** [comment]
Yes, I understand how this isn't a priority. Changes are slow. I get it.

---

> None of your forks are packaged by any major Linux distro (yet),

Chicken and egg situation? A project like Streamlink must start depending on it to get packagers to actually care. And unless mistaken, everyone (e.g. streamlink like projects) says exactly that.

> your urllib3.future fork also replaces the urllib3 namespace

Actually yes and no. OS packager are forbidden to replace the "legacy" namespace. It is documented and I try to make sure no accidental changes are pushed. And we could discuss this one further on for sure. You will see how we though this through and how we ensured no friction at all cost.

> Then there's also the lack of support for trio/anyio for the AsyncSession's runloop

If this was the only concern, I could add its support (via optional dependencies). But only if you were serious on choosing Niquests.

> Streamlink won't be able to switch to it, despite of all the problems it would solve and all the features it would offer.

We are constantly trying to listen to project like yours to improve ourselves. This could be a unique opportunity for us both.

Regards,

**Comment by Ousret** [comment]
tiny update: niquests is growing faster than I expected, we should expect more OSes to have it soon. it's around 60k pull a day today (solely PyPI).
httpx situation isn't improving so far.

you may just as well have it as an optional dependency to resolve the HTTP/1 shadowban some provider have (escape hatch for users) (and it will grow for sure)

regards,
